downloader/src/main/kotlin/Downloader.kt

```diff
@@ -77,7 +80,32 @@ object Downloader {
 
         val exception = DownloadException("Download failed for '${pkg.id.toCoordinates()}'.")
 
-        // Try downloading from VCS.
+        return config.sourcesOrigin.distinct().asSequence().mapNotNull { origin ->
+            handleDownload(origin, pkg, outputDirectory, allowMovingRevisions, exception)
+        }.firstOrNull() ?: throw exception
```

Comment:
Just as a suggestion, using `asSequence()` with `mapNotNull` and `firstOrNull` in an interesting approach, but I find the logic easier to understand if it is a bit more verbose:

```
config.sourcesOrigin.distinct().forEach { origin ->
    val provenance = handleDownload(...)
    if (provenance != null) return provenance
}

throw DownloadException(...)
```

Reply:
Yeah, I agree it's more obvious what's going on in @mnonnenmacher's version, although I tend to dislike return statements from within loops, as they can easily be overlooked. I'm fine with either implementation, though.

Reply:
Yes, I agree.
Is that change ok for you, @sschuberth?

Reply:
As stated above, either implementation is fine with me, as both have their pros and cons.